### DISCUSSION TITLE
Context / Wrapper dependent default ReaderOptions?

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#722] [Polls]
The Kotlin/Native wrapper PR discussion reminded me of an open question I had in my mind: What are the best default performance related `ReaderOptions` and should they be equal across all APIs or depend on the context, like mobile vs. desktop/backend?

For the general C++ interface, I came to the conclusion that the defaults are best set such that they maximize the detection rate and every deviation from that is basically to make the detection faster. In the [Android](https://github.com/zxing-cpp/zxing-cpp/blob/4c5c0732c10f96b7ca5948684d26bcd6a0a7be91/wrappers/android/zxingcpp/src/main/java/zxingcpp/BarcodeReader.kt#L66) wrapper the defaults are set to maximize speed. In [iOS](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/ios/Sources/Wrapper/Reader/ZXIReaderOptions.h) I don't even know what is happening. It seems there are not defaults at all, so one has to specify all settings explicitly? (The iOS demo does unfortunately not use any options setting).

There is the trivial consistency argument to make them behave the same by default. It also solves the question of how to deal with Kotlin/Native which can't be clearly put in either mobile or desktop. What is your preference? If you answer `NO`, please leave a comment explaining why.

**Comment by markusfisch** [comment]
iOS defaults to the C++ defaults, see [here](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/ios/Sources/Wrapper/Reader/ZXIReaderOptions.mm#L16).

**Reply by axxel** [reply]
Thanks. I understood as much. But I was surprised that [initWithFormats](https://github.com/zxing-cpp/zxing-cpp/blob/4c5c0732c10f96b7ca5948684d26bcd6a0a7be91/wrappers/ios/Sources/Wrapper/Reader/ZXIReaderOptions.h#L51) actually requires the caller to specify *all* parameters. Is that correct? If so, this function seems basically useless to me. I would assume someone would then always use `init` followed by individual calls of `setter` functions. If all that is correct, then the `initWithFormats` function should simply be removed. It also makes it basically impossible to deprecate the 3 settings that we recently discussed.

There has been 1 'No' vote and your comment is the only one so far, so you voted 'No'? If so why?

**Reply by markusfisch** [reply]
Yes, `initWithFormat` requires a caller to specifiy all parameters - something that doesn't sit right with me as well. I've checked with our iOS guys and they always use `init` and the appropriate `setter` functions, so I think it makes sense to remove `initWithFormat`. I've already filed a [PR](https://github.com/zxing-cpp/zxing-cpp/pull/723) for this.

> There has been 1 'No' vote and your comment is the only one so far, so you voted 'No'? If so why?

Yes, that was me 😉 But I'm not too sure about this. My initial thought was that some platforms might work better with special default settings. But it sure is desireable to have ZXingCpp work everywhere the same out of the box, too.

**Reply by axxel** [reply]
Awesome, one step closer to the ideal repository (the empty one, because it contains 0 bugs and is trivial to comprehend ;)).

@benjohnde your last chance to come up with a convincing argument why I should not merge this.
